civis/base.py

```diff
@@ -220,10 +224,27 @@ def _civis_state(self):
     @property
     def _state(self):
         """State of the CivisAsyncResultBase in `future` language."""
+        if self.__state is not None:
+            return self.__state
         with self._condition:
             return STATE_TRANS[self._civis_state]
 
     @_state.setter
     def _state(self, value):
         # Ignore attempts to set the _state from the `Future` superclass
-        pass
+        # TODO explain
+        self.__state = value
+
+    def set_result(self, result):
+        # TODO explain this
+        self.__state = futures._base.RUNNING
```

Comment:
This fiddling with the job state is pretty unpleasant, but I didn't see a clear alternative.

The challenge is that the state of a CivisFuture (or asyncresultbase, more specifically) is simply a [translation](https://github.com/civisanalytics/civis-python/blob/fd7700c80f1757b90f1f63a5728961e1d039c988/civis/base.py#L220-L224) of the Civis job state, which seems to violate an assumption in python's Futures that the state is only changed within the set_result() or set_exception() functions.

In 3.8, the set_result/set_exception functions [now check](https://github.com/python/cpython/blob/dec367261e7e2bb4dd42feeb58031abed2ade683/Lib/asyncio/futures.py#L236-L249) the state and make sure it's not already done before setting it to done.

Reply:
Why even have the override? Why not directly update `self._state`

Reply:
I'm a little confused. The link in your comment points to a `Futures` class that is [not the same](https://github.com/python/cpython/blob/dec367261e7e2bb4dd42feeb58031abed2ade683/Lib/asyncio/futures.py#L30) as `concurrent.futures.Future`. 

Is this what you meant: https://github.com/python/cpython/blob/3.8/Lib/concurrent/futures/_base.py#L523

Reply:
ah, yup. sorry about the bad link.  i updated my original comment.

Reply:
> Why not directly update `self._state`

`civis.base.AsyncResultBase._state` is actually [a function with a property decorator](https://github.com/civisanalytics/civis-python/blob/bb9c6df4023f96897ec728ce4251b2016322ad6d/civis/base.py#L226-L229) that ignores attempts to set it, the idea being that platform provides the actual state of the future object. I think this is probably undesirable, but I didn't see a simple way to unwind that assumption without a substantial refactor, which doesn't seem warranted here since it has worked pretty reliably for a long time.